common/events/events_io.cpp

```diff
@@ -30,12 +30,16 @@ AsyncReaderFromReader::~AsyncReaderFromReader() {
 
 error::Error AsyncReaderFromReader::AsyncRead(
 	vector<uint8_t>::iterator start, vector<uint8_t>::iterator end, mio::AsyncIoHandler handler) {
-	// Simple, "cheating" implementation, we just do it synchronously.
-	in_progress_ = true;
-	loop_.Post([this, start, end, handler]() {
-		auto result = reader_->Read(start, end);
-		in_progress_ = false;
-		handler(result);
+	cancelled_ = make_shared<bool>(false);
+	auto cancelled = cancelled_;
+	loop_.Post([this, cancelled, start, end, handler]() {
```

Comment:
Wouldn't `cancelled_` (i.e. captured as a copy) be the same here?

Reply:
That would make perfect sense, but C++ doesn't allow you to capture member variables directly, you have to make them into local variables first.

However, I realize now that I can make the local variable a reference, so that I avoid at least one copy (still one in the capture, but that's intentional).